core/src/main/java/org/frankframework/processors/InputOutputPipeProcessor.java

```diff
@@ -118,23 +118,33 @@ protected PipeRunResult processPipe(@Nonnull PipeLine pipeLine, @Nonnull IPipe p
 				log.debug("Pipeline of adapter [{}] restoring from compacted result for pipe [{}]", owner::getName, pipe::getName);
 				Message result = pipeRunResult.getResult();
 				if (!Message.isEmpty(result)) {
+					InputSource inputSource;
 					try {
-						String resultString = result.asString();
-						pipeRunResult.setResult(RestoreMovedElementsHandler.process(resultString, pipeLineSession));
+						// Preserve the message so that it can be read again, in case there was an error during restoring
+						result.preserve();
+						inputSource = result.asInputSource();
 					} catch (IOException e) {
-						throw new PipeRunException(pipe, "cannot open stream of result", e);
+						throw new PipeRunException(pipe, "Pipeline of [" + pipeLine.getOwner().getName() + "] could not read received message during restoring of moved elements: " + e.getMessage(), e);
+					}
+					XmlWriter xmlWriter = new XmlWriter();
+					RestoreMovedElementsHandler handler = new RestoreMovedElementsHandler(xmlWriter);
+					handler.setSession(pipeLineSession);
+					try {
+						XmlUtils.parseXml(inputSource, handler);
+						pipeRunResult.setResult(xmlWriter.getWriter());
+					} catch (Exception e) {
+						log.warn("Pipeline of adapter [{}] could not restore moved elements on the received message: {}", owner.getName(), e.getMessage());
 					}
 				}
```

Comment:
It would be nice to have this code extracted into a separate method, to clean up the main method and make the flow of that more readable.